### PR TITLE
Implement hot exit on switching folders

### LINKED ISSUE #15467: Hot exit on switching folders
Parent issue: #101

Sublime has a feature that allows hot exit to *always* be allowed on a window provided it has a project associated with it. The initial concern still remains that backups may go missing if we just always enable this, the Sublime project system is nice because the act of creating a project (`.sublime-project`) is basically an opt in for project-level features.

Interested in people's ideas on this, here are a few:

- Enabling switching folders all the time?
- Enable if `<project>/.vscode` exists?
- Enable if a setting is toggled?
- Enable if a setting includes the workspace path? (has issues with multi-folder workspaces)

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes #15467, see the issue for detailed discussion on how it works.

I added tests for the hot exit window close case for single window/app exit and multiple window.

### CODE REVIEW

File: src/vs/workbench/parts/files/browser/files.contribution.ts [hunk 1]
```diff
@@ -244,9 +244,10 @@ configurationRegistry.registerConfiguration({
...
 		},
 		'files.hotExit': {
-			'type': 'boolean',
-			'default': true,
-			'description': nls.localize('hotExit', "Controls whether unsaved files are restored after relaunching. If this is enabled there will be no prompt to save when exiting the editor.")
+			'type': 'string',
+			'enum': [HotExitConfiguration.OFF, HotExitConfiguration.ON_EXIT, HotExitConfiguration.ON_EXIT_AND_WINDOW_CLOSE],
+			'default': HotExitConfiguration.ON_EXIT,
+			'description': nls.localize('hotExit', "Whether hot exit is enabled which allows changes to unsaved files to be remembered between sessions, hiding the prompt t save when exiting the editor. Selecting \"{0}\" means that hot exit will only be triggered when the application is closed (workbench.action.quit command via command pallete, keybinding of menu) and ALL windows with backups will be restored upon next launch. Selecting \"{1}\" will trigger hot exit when any FOLDER window is closed, only NON-FOLDER windows will be restored when the application is restarted (not FOLDER workspaces).", HotExitConfiguration.ON_EXIT, HotExitConfiguration.ON_EXIT_AND_WINDOW_CLOSE)
```
Comment (bpasero) [inline src/vs/workbench/parts/files/browser/files.contribution.ts]: Typo?

`hiding the prompt t save`

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
@Tyriar two things:

First of all, I am getting a confirmation dialog when I have dirty files inside a folder workspace and I pick another folder to open from File > Open Recent. I would not expect that according to the discussion we had.

Second, I am a bit confused about this behaviour: Open 2 windows with folders and have dirty files in each. Do File > Quit and restart. You end up with only one window restored. Should we not restore both windows? My understanding is that this setting is for closing a single window and allowing hot exit if that window has a folder opened. But when I close multiple windows with dirty files we could still restore all. Or is the idea to let this be controlled via the `window.reopenFolders` setting?

**Comment by Tyriar** [comment]
Looks like I missed the LOAD shutdown reason. I'll look at that and the tests tomorrow. 

Yes the idea is to rely on `window.reopenFolders` in that case to give the user more control more control.

File: src/vs/workbench/parts/files/browser/files.contribution.ts (hunk 1 above)
Comment (Tyriar) [inline src/vs/workbench/parts/files/browser/files.contribution.ts]: https://github.com/Microsoft/vscode/pull/18626/commits/749f99aa893d58b1799ca593e99b63937cb0f780

**Comment by Tyriar** [comment]
@bpasero please have another look. I built out the tests to have every shutdown case covered.

**Comment by bpasero** [comment]
@Tyriar LGTM

**Review by bpasero — APPROVED** [review]

**Comment by bpasero** [comment]
@Tyriar looks like there are test failures though on Linux in travis CI

**Comment by Tyriar** [comment]
I think that should fix it, will merge tomorrow if all pass :+1:

**Comment by Tyriar** [comment]
I think that should fix it, will merge when confirmed :+1:

**Pull request merged**